firedrake/preconditioners/offload.py

```diff
@@ -0,0 +1,118 @@
+from firedrake.preconditioners.base import PCBase
+from firedrake.functionspace import FunctionSpace, MixedFunctionSpace
+from firedrake.petsc import PETSc
+from firedrake.ufl_expr import TestFunction, TrialFunction
+import firedrake.dmhooks as dmhooks
+from firedrake.dmhooks import get_function_space
+
+__all__ = ("OffloadPC",)
+
+
+class OffloadPC(PCBase):
```

Comment:
Could `AssembledPC` assume this functionality by providing `-assembled_mat_type aijcusparse`?

Reply:
We could have this one as a subclass of `AssembledPC`, there's substantial code duplication

Reply:
That seems like a good idea. @Olender

Reply:
Good idea! I implemented this in the latest commit, but I'm still testing things out to ensure everything works as expected. Let me know if you have any more suggestions